util/secscan/blob.py

```diff
@@ -0,0 +1,75 @@
+from urlparse import urljoin
+
+from flask import url_for
+
+from util.security.registry_jwt import generate_bearer_token, build_context_and_subject
+
+
+def _repository_and_namespace(repository_ref):
+    repo_name = repository_ref.name
+    namespace_name = repository_ref.namespace_name
+    return "/".join([namespace_name, repo_name])
+
+
+class BlobURLRetriever(object):
+    """
+    Helper class which encodes the logic for retrieving blobs for security indexing.
+    """
+
+    def __init__(self, storage, instance_keys, app):
+        assert storage is not None
+        assert instance_keys is not None
+        assert app is not None
+
+        self._storage = storage
+        self._instance_keys = instance_keys
+        self._app = app
+
+    def url_for_download(self, repository_ref, blob):
+        """
+        Returns the URL for downloading the given blob under the given repository.
+        """
+        # Try via direct download from storage.
+        uri = self._storage.get_direct_download_url(self._storage.locations, blob.storage_path)
+        if uri is not None:
+            return uri
+
+        # Otherwise, use the registry blob download endpoint.
+        with self._app.test_request_context("/"):
+            relative_layer_url = url_for(
+                "v2.download_blob",
+                repository=_repository_and_namespace(repository_ref),
+                digest=blob.digest,
+            )
+
+        url_scheme_and_hostname = "%s://%s" % (
+            self._app.config["PREFERRED_URL_SCHEME"],
+            self._app.config["SERVER_HOSTNAME"],
+        )
+        return urljoin(url_scheme_and_hostname, relative_layer_url)
+
+    def headers_for_download(self, repository_ref, blob, timeout=60):
+        """
+        Returns the headers for downloading the given blob under the given repository.
+        """
+        uri = self._storage.get_direct_download_url(self._storage.locations, blob.storage_path)
+        if uri is not None:
+            return {}
+
+        # Otherwise, we mint a JWT and place it into an Auth header.
+        audience = self._app.config["SERVER_HOSTNAME"]
+        context, subject = build_context_and_subject()
+        access = [
+            {
+                "type": "repository",
+                "name": _repository_and_namespace(repository_ref),
+                "actions": ["pull"],
+            }
+        ]
+        assert set(act for acs in access for act in acs["actions"]) == {"pull"}
+
+        auth_token = generate_bearer_token(
+            audience, subject, context, access, timeout, self._instance_keys
+        )
+
+        return {"Authorization": ["Bearer " + auth_token]}
```

Comment:
Clair reports this error:
```
{"level":"debug","component":"init/NewHttpTransport","component":"libindex/Libindex.Index","manifest":"sha256:02d6a5382a93614536adeacb7a4fd39eeebeeb65ebc3f2a4bbef345bfd14e7f4","component":"libindex/Libindex.index","component":"internal/indexer/controller/Controller.Index","manifest":"sha256:02d6a5382a93614536adeacb7a4fd39eeebeeb65ebc3f2a4bbef345bfd14e7f4","state":"CheckManifest","state":"FetchLayers","component":"internal/indexer/fetcher/fetcher.fetch","layer":"sha256:50495ac9ea81fc20716c0547b58cebadf3631c39fd0b242129af78d05aa9640b","uri":"https://registry.skynet-dev.com/v2/alecmerdler/bad-image/blobs/sha256:50495ac9ea81fc20716c0547b58cebadf3631c39fd0b242129af78d05aa9640b","content-type":"text/html; charset=utf-8","time":"2020-05-14T23:28:51Z","time":"2020-05-14T23:28:51Z","message":"reported content-type"}
{"level":"warn","component":"init/NewHttpTransport","component":"libindex/Libindex.Index","manifest":"sha256:02d6a5382a93614536adeacb7a4fd39eeebeeb65ebc3f2a4bbef345bfd14e7f4","component":"libindex/Libindex.index","component":"internal/indexer/controller/Controller.Index","manifest":"sha256:02d6a5382a93614536adeacb7a4fd39eeebeeb65ebc3f2a4bbef345bfd14e7f4","state":"CheckManifest","state":"FetchLayers","error":"encountered error while fetching a layer: fetcher: unknown content-type \"text/html; charset=utf-8\"","time":"2020-05-14T23:28:51Z","time":"2020-05-14T23:28:51Z","message":"layers fetch failure"}
```

I investigated our `v2.download_blob` route, and saw we were [are setting the correct](https://github.com/quay/quay/blob/master/endpoints/v2/blob.py#L50) `Content-Type` header. However, trying it manually using cURL reveals our true headers due to something in NGINX:
```
$ curl -sSL -D - https://registry.skynet-dev.com/v2/alecmerdler/bad-image/blobs/sha256:50495ac9ea81fc20716c0547b58cebadf3631c39fd0b242129af78d05aa9640b -o /dev/null
HTTP/2 200 
server: nginx/1.12.1
date: Fri, 15 May 2020 03:35:01 GMT
content-type: text/html; charset=utf-8
cache-control: max-age=31536000
x-frame-options: DENY
strict-transport-security: max-age=63072000; preload
```

Reply:
Occurs with `HTTP/1.1` as well:
```
$ curl --http1.1 -sSL -D - https://registry.skynet-dev.com/v2/alecmerdler/bad-image/blobs/sha256:50495ac9ea81fc20716c0547b58cebadf3631c39fd0b242129af78d05aa9640b -o /dev/null
HTTP/1.1 200 OK
Server: nginx/1.12.1
Date: Fri, 15 May 2020 07:36:38 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: close
Cache-Control: max-age=31536000
X-Frame-Options: DENY
Strict-Transport-Security: max-age=63072000; preload
```

Reply:
Doesn't occur on Quay.io (not using local storage):
```
$ curl -sSL -D - https://quay.io/v2/alecmerdler/bad-image/blobs/sha256:e5dd9dbb37df5b731a6688fa49f4003359f6f126958c9c928f937bec69836320 -o /dev/null
HTTP/2 404 
server: nginx/1.12.1
date: Fri, 15 May 2020 07:47:27 GMT
content-type: application/json
content-length: 86
```

Reply:
It arguably still does, the 301 that quay.io issues still has `text/html` as the content-type, it's just not used because it's a redirect

Reply:
> I investigated our `v2.download_blob` route, and saw we were [are setting the correct](https://github.com/quay/quay/blob/master/endpoints/v2/blob.py#L50) `Content-Type` header.

This seems to only be happening on HEAD?

Reply:
Problem was because we were setting the headers to the return value of `dict.update`, which is `None`. Fixed in another change on this PR.